tests/gpflow/posteriors/test_posteriors.py

```diff
@@ -0,0 +1,445 @@
+_linear_coregionalization = [LinearCoregionalizationPosterior]
+
+
+def test_no_missing_kernels():
+    tested_kernels = set(
+        chain(
+            _independent_single_output,
+            _fully_correlated_multi_output,
+            _independent_multi_output,
+            _fallback_independent_multi_output,
+            _linear_coregionalization,
+        )
+    )
+
+    available_kernels = list(gpflow.ci_utils.subclasses(AbstractPosterior))
+    concrete_kernels = set([k for k in available_kernels if not isabstract(k)])
+
+    assert tested_kernels == concrete_kernels
+
+
+@dataclass(frozen=True)
+class LatentVariationalParameters:
+    qsqrt = (np.random.randn(NUM_INDUCING_POINTS, 1) ** 2) * 0.01
+    qmean = np.random.randn(NUM_INDUCING_POINTS, 1)
+
+
+def _assert_fused_predict_f_equals_precomputed_predict_f(
+    posterior, precompute, full_cov, full_output_cov, decimals=6
+):
+    Xnew = np.random.randn(20, 1)
+
+    fused_f_mean, fused_f_cov = posterior.fused_predict_f(
+        Xnew, full_cov=full_cov, full_output_cov=full_output_cov
+    )
+
+    if not precompute:
+        posterior.update_cache()
+
+    precomputed_f_mean, precomputed_f_cov = posterior.predict_f(
+        Xnew, full_cov=full_cov, full_output_cov=full_output_cov
+    )
+
+    np.testing.assert_array_almost_equal(fused_f_mean, precomputed_f_mean, decimal=decimals)
+    np.testing.assert_array_almost_equal(fused_f_cov, precomputed_f_cov, decimal=decimals)
+
+
+@pytest.mark.parametrize("posterior_class", _independent_single_output)
+def test_independent_single_output(
+    posterior_class, whiten, mean_function, precompute, full_cov, full_output_cov
+):
+    kernel = gpflow.kernels.SquaredExponential()
+    inducing_variable = inducingpoint_wrapper(np.random.randn(NUM_INDUCING_POINTS, 1))
+
+    posterior = posterior_class(
+        kernel=kernel,
+        inducing_variable=inducing_variable,
+        q_mu=LatentVariationalParameters.qmean,
+        q_sqrt=tf.constant(LatentVariationalParameters.qsqrt),
+        whiten=whiten,
+        mean_function=mean_function,
+        precompute=precompute,
+    )
+
+    _assert_fused_predict_f_equals_precomputed_predict_f(
+        posterior, precompute, full_cov, full_output_cov
+    )
+
+
+@dataclass(frozen=True)
+class LatentVariationalMultiOutputParameters:
+    qsqrt = np.eye(NUM_INDUCING_POINTS)[None, :]
+    qmean = np.random.randn(NUM_INDUCING_POINTS, 1)
+
+
+@pytest.mark.parametrize("posterior_class", _fully_correlated_multi_output)
+@pytest.mark.parametrize(
+    "q_sqrt", [None, tf.constant(LatentVariationalMultiOutputParameters.qsqrt)]
+)
+@pytest.mark.parametrize(
+    "whiten",
+    [
+        True,
+        pytest.param(
+            False,
+            marks=pytest.mark.xfail(
+                reason="FullyCorrelatedPosterior and subclasses are not consistent between the "
+                "fused and precomputed implementations with whiten=False."
```

Comment:
as in, they both produce an answer, but it differs?

Reply:
Yes, the answers from the two methods are different. I have not yet been able to find out why.

Reply:
I have raised #1662 and #1663 to fix these issues.